src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs

```diff
@@ -314,20 +321,37 @@ public void Initialize(IConsole console)
                         UpdateWorkingDirectory();
                         await ExecuteInitScriptsAsync();
 
-                        // check if PMC console is actually opened, then only hook to solution load/close events.
+                        // check if PMC console is actually opened, then only hook to solution load events.
                         if (console is IWpfConsole)
                         {
                             // Hook up solution events
                             _solutionManager.Value.SolutionOpened += (_, __) => HandleSolutionOpened();
-                            _solutionManager.Value.SolutionClosed += (o, e) =>
+                        }
+
+                        _solutionManager.Value.SolutionClosed += (o, e) =>
```

Comment:
I was expecting a meeting where we could discuss the telemetry at a high level, rather than diving straight into the implementation immediately.

NuGet starts the powershell host on first usage, whether that's a packages.config package install that uses install.ps1, or opening PMC. The PowerShell host then runs for the rest of the devenv process, regardless of the solutions.

Therefore, what happens when:
* A customer starts VS, does not load a solution, and uses PMC, closes VS. Will we get any telemetry?
* A customer starts VS, loads a solution, does not use PowerShell, close the solution, open second solution in the same process, opens PMC.
* A customer starts VS, loads a solution, loads PMC, closes PMC, opens a second solution and does not use PMC.

In the last two scenarios, the powershell host is in memory, but if we query telemetry, it will look like the powershell host was only loaded 50% of the time, whereas it was actually loaded in 75% of the solution sessions, and 100% of the VS sessions.

Furthermore, you're adding this telemetry in NuGetConsole's PowerShellHost. Does't this mean that a devenv process that never uses NuGet's powershell host will not generate any telemetry?  How will we determine the percentage of VS sessions that use powershell, without telemetry events that say "feature was not used"?

Your PR says
> I am not trying to answer all questions now or propose all the possible questions/hypothesis without even basic telemetry. Let's just have minimum telemetry for PMC and init1.ps1 scripts and commands run on PMC. Then after looking at that one let's decide if we need more telemetry. Problem is currently we don't have any telemetry for above.

The problem I was trying to state in our previous meetings is that if we don't figure out what questions we want to answer, then the telemetry we add might be useless. If we don't consider the question first, then we implement telemetry, like this PR is proposing, then we have to wait until hotseat inserts, then we have to wait for the version to ship, then we start getting data. If we query the data and find it doesn't help us make decisions, we've wasted weeks or months. This is exactly the problem I had with the package source telemetry I added nearly a year ago, so I know from first-hand experience. Once we started getting data and tried to query it, it didn't give us useful information to make decisions.

Reply:
> Therefore, what happens when:
> 
> * A customer starts VS, does not load a solution, and uses PMC, closes VS. Will we get any telemetry?

We still get [PackageManagerConsoleLoadCount](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L89), [ReOpenAtStart](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGet.Console/Xamls/ConsoleContainer.xaml.cs#L98) events this case, and I imagine it hardly ever happen and but we'll know when it happens. But we don't get [PowerShellExecuteCommand](https://github.com/NuGet/NuGet.Client/blob/9f84b339bc8abf1dcdb2e4f975ac2a1a5865c8dd/src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs#L334) event this case since there is no solution.

> * A customer starts VS, loads a solution, does not use PowerShell, close the solution, open second solution in the same process, opens PMC.
> * A customer starts VS, loads a solution, loads PMC, closes PMC, opens a second solution and does not use PMC.

>In the last two scenarios, the powershell host is in memory, but if we query telemetry, it will look like the powershell host was only loaded 50% of the time, whereas it was actually loaded in 75% of the solution sessions, and 100% of the VS sessions.

I'm not aware of use case for how many % of solution session duration had poweshell host is loaded, we're more interested how many times it's loaded without user's intention of use. If `powershell is loaded without users intention of` use less than 0.1% (which I believe) then there is not much to dig more since it's very rare case, but if it happen 5%-10% of all VS open session then above question is more interesting. 

> Furthermore, you're adding this telemetry in NuGetConsole's PowerShellHost. Does't this mean that a devenv process that never uses NuGet's powershell host will not generate any telemetry? How will we determine the percentage of VS sessions that use powershell, without telemetry events that say "feature was not used"?

VS itself already generate ton of telemetry, no need to add another telemetry for `feature not uses`. We can aggregate total number of sessions from Kusto table. Current new telemetry is for `powershell loaded case`. So `powershell %` =   `powershell loaded case`/ `Total sessions`.

>The problem I was trying to state in our previous meetings is that if we don't figure out what questions we want to answer, then the telemetry we add might be useless. If we don't consider the question first, then we implement telemetry, like this PR is proposing, then we have to wait until hotseat inserts, then we have to wait for the version to ship, then we start getting data. If we query the data and find it doesn't help us make decisions, we've wasted weeks or months. This is exactly the problem I had with the package source telemetry I added nearly a year ago, so I know from first-hand experience. Once we started getting data and tried to query it, it didn't give us useful information to make decisions.

Your concern is valid. We had discussion and asked for inputs on teams, in the end conclusion was up to me come up with what to do about this. I believe this implementation will help us answer 5 questions posted on description body of this PR. 
Based on what we get from telemetry then we'll decide if we need more telemetry or going to have more questions need answers. We can't come up with all the questions we're trying to answer once without any preliminary basic data.
For example `ReOpenAtStart` events happens less 1.000 times in 1 million then it's pretty much self explanatory, it's very rare case and no need future action. If it happens 40.000 times in 1 million then we need to answer more questions.

I acknowledge this could be not enough, but it's good start. Please comment if you have any more input.

Reply:
> VS itself already generate ton of telemetry, no need to add another telemetry for feature was not used.

It's a trade-off of usefulness vs quantity of events. We generate multiple events per restore (we really should merge them into 1 event with multiple measurements/properties, but that's a different issue), and we do one restore every build, so many per typical solution session. Hopefully we find that information useful enough to justify the quantity of data.

How valuable do you think a chart saying "for each version of VS, x% used PM UI at least once. y% used PMC at least once. z% have the 'Ask on first install which package format to use' setting enabled". I think that would be useful to know about what features customers actually care about, would be interesting for us to understand how it changes over time and across different versions of Visual Studio, and would allow us to make better decisions about what work to priotitise.  A single "NuGetFeaturesUsed" event per VS session with multiple properties could achieve this. So, I disagree that adding telemetry about features not used will lead to unreasonable telemetry volume.


> We had discussion and asked for inputs on teams, in the end conclusion was up to me come up with what to do about this.

Yes, I was part of that meeting. But I wasn't invited to a follow up meeting to discuss the proposal. As far as I can tell, you went straight to an implementation without verifying with the team that these are the things we want to collect. Maybe I'm wrong, if it was discussed and other people agree this is a good approach, then that's good.

I don't know what the pros and cons are for collecting events per VS session vs solution session, or why the decision was made to do it the way it is implemented in this PR. I don't know if I should be giving feedback only on the "high level approach", or also on the line-by-line review like PRs I expect to get merged.

Reply:
Currently I don't know how to emit event by default when VS close from nuget, if you know please guide me for how to do it. I don't have any previous experience with `IVs` related things.
But I found where to hook when [VS close](https://github.com/NuGet/NuGet.Client/blob/d287e729bbcd6aa404c9ed26184a4d373a9e97cf/src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs#L1113). Still I don't know how to pass telemetry info between different assemblies. 
Currently I don't see any example I can use of.
To make clear things I scheduled meeting for tomorrow.

Reply:
Also one of reason I went into implementation is I need to learn first what is possible and what is not possible with VS/Nuget. 
Without any such concept it's hard to come up good questions. Now I have bit more understanding of how it works. Let's discuss tomorrow in details.